File: app/src/main/java/com/iitk/proxyLogin/SummaryNotification.java
package com.iitk.proxyLogin;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.graphics.Bitmap;
import android.util.Log;

import java.util.ArrayList;
import java.util.Iterator;

import static android.app.Notification.*;

/**
 * Created by kshivang on 12/04/17.
 *
 */

public class SummaryNotification {
    private static final String TAG;
    private static int notificationId;
    private boolean alert;
    Context context;
    ArrayList<String> lines;
    Builder mBuilder;
    private String message;
    private Bitmap picture;
    private String summaryMessage;
    private String tag;

    static {
        TAG = SummaryNotification.class.getSimpleName();
        notificationId = 1001;
    }

    public SummaryNotification(Context context, String title, String message, String ticker,
                               int iconResource, String tag) {
        this.alert = false;
        this.picture = null;
        this.tag = tag;
        this.context = context;
        this.message = message;
        this.lines = new ArrayList();
        this.mBuilder = new Builder(context).setContentTitle(title)
                .setContentText(message).setSmallIcon(iconResource)
                .setTicker(ticker).setAutoCancel(true);
        if (message != null) {
            this.lines.add(message);
        }
    }

    public void setContentIntent(PendingIntent intent) {
        this.mBuilder.setContentIntent(intent);
    }

    public void addAction(int iconResource, String actionName, PendingIntent intent) {
        this.mBuilder.addAction(iconResource, actionName, intent);
    }

    public void updateTitle(String title) {
        this.mBuilder.setContentTitle(title);
    }

    public void setNotificationId(int notificationId) {
        notificationId = notificationId;
    }

    public void show() {
        if (this.picture == null) {
            InboxStyle inboxStyle = new InboxStyle();
            Iterator it = this.lines.iterator();
            while (it.hasNext()) {
                inboxStyle.addLine((String) it.next());
            }
            if (this.summaryMessage != null) {
                inboxStyle.setSummaryText(this.summaryMessage);
            }
            this.mBuilder.setStyle(inboxStyle);
        } else {
            BigPictureStyle bigPictureStyle = new BigPictureStyle();
            bigPictureStyle.bigPicture(this.picture);
            if (this.summaryMessage != null) {
                bigPictureStyle.setSummaryText(this.summaryMessage);
            }
            this.mBuilder.setStyle(bigPictureStyle);
        }
        Notification notification = null;
        try {
            notification = this.mBuilder.build();
        } catch (NullPointerException ex) {
            Log.e(TAG, "Got NPE inboxStyle lines " + this.lines, ex);
        }
        if (notification != null) {
            if (this.alert) {
                notification.defaults |= -1;
            }
            ((NotificationManager) this.context.getSystemService(Context.NOTIFICATION_SERVICE)).notify(this.tag, notificationId, notification);
        }
    }

    public static void cancelNotification(Context context, String tag, int id) {
        ((NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE)).cancel(tag, id);
    }

    public void setLargeIcon(Bitmap icon) {
        this.mBuilder.setLargeIcon(icon);
    }

    public String getMessage() {
        return this.message;
    }

    public void setMessage(String message) {
        this.mBuilder.setContentText(message);
    }

    public void setProgress(int progress, int max) {
        this.mBuilder.setProgress(max, progress, false);
    }

    public void addLineToBigView(String line) {
        this.lines.add(line);
    }

    public void setSummaryMessage(String summaryMessage) {
        this.summaryMessage = summaryMessage;
    }

    public void setContentInfo(String contentInfo) {
        this.mBuilder.setContentInfo(contentInfo);
    }

    public void setAlert(boolean alert) {
        this.alert = alert;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    public void setPicture(Bitmap picture) {
        this.picture = picture;
    }
}

File: app/src/main/java/com/iitk/proxyLogin/notfication/SummaryNotification.kt
package com.iitk.proxyLogin.notfication

import android.app.Notification
import android.app.Notification.*
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import java.util.*

/**
 * Created by kshivang on 12/04/17.

 */

class SummaryNotification(var context: Context, title: String, private val message: String?, ticker: String,
                          iconResource: Int, private var tag: String?) {
    private var alert: Boolean = false
    var lines: ArrayList<String>
    var mBuilder: Builder
    private var picture: Bitmap? = null
    private var summaryMessage: String? = null

    init {
        this.alert = false
        this.picture = null
        this.lines = ArrayList()
        this.mBuilder = Builder(context).setContentTitle(title)
                .setContentText(message).setSmallIcon(iconResource)
                .setTicker(ticker).setAutoCancel(true)
        if (message != null) {
            this.lines.add(message)
        }
    }

    fun setContentIntent(intent: PendingIntent) {
        this.mBuilder.setContentIntent(intent)
    }

    fun addAction(iconResource: Int, actionName: String, intent: PendingIntent) {
        this.mBuilder.addAction(iconResource, actionName, intent)
    }

    fun updateTitle(title: String) {
        this.mBuilder.setContentTitle(title)
    }

    fun setNotificationId(notificationId: Int) {
        var notificationId = notificationId
        notificationId = notificationId
    }

    fun show() {
        if (this.picture == null) {
            val inboxStyle = InboxStyle()
            val it = this.lines.iterator()
            while (it.hasNext()) {
                inboxStyle.addLine(it.next())
            }
            if (this.summaryMessage != null) {
                inboxStyle.setSummaryText(this.summaryMessage)
            }
            this.mBuilder.setStyle(inboxStyle)
        } else {
            val bigPictureStyle = BigPictureStyle()
            bigPictureStyle.bigPicture(this.picture)
            if (this.summaryMessage != null) {
                bigPictureStyle.setSummaryText(this.summaryMessage)
            }
            this.mBuilder.setStyle(bigPictureStyle)
        }
        var notification: Notification? = null
        try {
            notification = this.mBuilder.build()
        } catch (ex: NullPointerException) {
            Log.e(TAG, "Got NPE inboxStyle lines " + this.lines, ex)
        }

        if (notification != null) {
            if (this.alert) {
                notification.defaults = notification.defaults or -1
            }
            (this.context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).notify(this.tag, notificationId, notification)
        }
    }

    fun setLargeIcon(icon: Bitmap) {
        this.mBuilder.setLargeIcon(icon)
    }

    fun getMessage(): String? {
        return this.message
    }

    fun setMessage(message: String) {
        this.mBuilder.setContentText(message)
    }

    fun setProgress(progress: Int, max: Int) {
        this.mBuilder.setProgress(max, progress, false)
    }

    fun addLineToBigView(line: String) {
        this.lines.add(line)
    }

    fun setSummaryMessage(summaryMessage: String) {
        this.summaryMessage = summaryMessage
    }

    fun setContentInfo(contentInfo: String) {
        this.mBuilder.setContentInfo(contentInfo)
    }

    fun setAlert(alert: Boolean) {
        this.alert = alert
    }

    fun setTag(tag: String) {
        this.tag = tag
    }

    fun setPicture(picture: Bitmap) {
        this.picture = picture
    }

    companion object {
        private val TAG: String
        private var notificationId: Int = 0

        init {
            TAG = SummaryNotification::class.java.simpleName
            notificationId = 1001
        }

        fun cancelNotification(context: Context, tag: String, id: Int) {
            (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).cancel(tag, id)
        }
    }
}
